### PR TITLE
Add csvzen-core: zero-alloc streaming CSV writer

### PR DESCRIPTION
**Pull request by guizmaii**
## Summary

- New `csvzen-core` module: a tiny Scala 3 library for writing RFC 4180-compliant CSV to a file in a mostly-zero-allocation, streaming fashion.
- Public API: `CsvConfig`, `CsvWriter.open(path, config)`, `FieldEmitter` (typed primitives straight to `java.io.Writer`), `CsvFieldEncoder[A]` (SAM typeclass with givens for `String`, `Int`, `Long`, `Short`, `Byte`, `Boolean`, `Char`, `Float`, `Double`, `BigInt`, `BigDecimal`, `UUID`, `Currency`, every `java.time.*` type, and `Option[A]`), `CsvRowEncoder[A]` with `Mirror.ProductOf` derivation for flat case classes (`derives CsvRowEncoder`). Headers are derived from field labels.
- Multi-module sbt build: root aggregate with a single `core` subproject today, structured so a future `csvzen-zio` sibling is a drop-in.
- Exhaustive test suite (117 tests): primitive edge cases + seeded randomised sweeps, full `Short`/`Byte` ranges, RFC 4180 escape matrix across configs, derivation for arities 1/2/5/10/22 (+ unicode / backticked labels), real-file round-trips through a test-only CSV reader for every shipped encoder, UTF-8 byte sequences for emoji/CJK, and negative compile-time tests for nested/non-product types.

## Design

See [`docs/superpowers/specs/2026-04-24-csvzen-design.md`](./docs/superpowers/specs/2026-04-24-csvzen-design.md).

Positioned against scala-csv (requires `Seq[String]` per row, `.toString` everywhere), kantan-csv (heavy implicits), and zio-blocks' `schema-csv` (codec concept good, but `.toString` per cell + per-row `StringBuilder`s). csvzen keeps the codec concept and strips everything else: `FieldEmitter` writes digits straight to the `Writer` for `Int`/`Long`/`Short`/`Byte`, with a documented one-alloc carve-out for `Float`/`Double`/`BigInt`/`BigDecimal`/`UUID`/`Currency`/`java.time.*`.

Approach A (SAM trait + JIT-inlining) chosen for v1; approach B (full-`inline` derivation) noted as a future experimental branch.

## Test plan

- [x] `sbt --client core/compile` clean
- [x] `sbt --client core/test` — 117 tests pass
- [x] 100k-row round-trip via real file
- [x] UTF-8 / no BOM / `\r\n` vs `\n` byte-level assertions
- [x] Negative derivation tests (nested case class, non-product)

### CODE REVIEW

File: csvzen/core/src/test/scala/com/guizmaii/csvzen/core/CsvFieldEncoderSpec.scala [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        val u1 = UUID.randomUUID()
+        val u2 = UUID.nameUUIDFromBytes("csvzen".getBytes("UTF-8"))
+        assertTrue(encodeOne(u1) == u1.toString, encodeOne(u2) == u2.toString)
+      },
+      test("Currency codes") {
+        Seq("EUR", "USD", "JPY").foreach { code =>
+          val c = Currency.getInstance(code)
+          assert(encodeOne(c))(Assertion.equalTo(code))
+        }
+        assertTrue(true)
```
Comment (guizmaii) [inline csvzen/core/src/test/scala/com/guizmaii/csvzen/core/CsvFieldEncoderSpec.scala]: **Test bug — assertions inside `.foreach` are discarded.** The `assert(...)(Assertion.equalTo(code))` inside the loop returns a `TestResult`, but `.foreach` ignores the return value, and the outer `assertTrue(true)` is what the test actually reports. If any currency code fails to encode, this test still passes green.

Suggested fix:
```scala
assertTrue(Seq("EUR", "USD", "JPY").forall(code => encodeOne(Currency.getInstance(code)) == code))
```

File: csvzen/core/src/main/scala/com/guizmaii/csvzen/core/CsvRowEncoder.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package com.guizmaii.csvzen.core
+
+import scala.collection.immutable.ArraySeq
```
Comment (guizmaii) [inline csvzen/core/src/main/scala/com/guizmaii/csvzen/core/CsvRowEncoder.scala]: Unused import — `ArraySeq` is only used in `DerivedCsvRowEncoder.scala`, not in this file. Safe to drop. (`-Wunused:imports` would catch this; consider enabling it alongside the existing `-Werror`.)

File: csvzen/core/src/test/scala/com/guizmaii/csvzen/core/CsvRowEncoderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    suite("derivation fails cleanly for unsupported shapes")(
+      test("missing CsvFieldEncoder for a field type") {
+        final class Custom(val x: Int)
+        // Custom has no CsvFieldEncoder — derivation on a case class carrying it fails.
+        val err = scala.compiletime.testing.typeCheckErrors(
+          """
+          final case class BadRow(c: Custom) derives CsvRowEncoder
+          """
+        )
+        assertTrue(err.nonEmpty)
```
Comment (guizmaii) [inline csvzen/core/src/test/scala/com/guizmaii/csvzen/core/CsvRowEncoderSpec.scala]: **Negative compile tests are under-specified.** `assertTrue(err.nonEmpty)` passes for *any* compile failure — a typo in the quoted snippet, a rename of `CsvRowEncoder`, or an unrelated import issue would all make the test green while no longer exercising the intended failure mode.

Pin an error message substring so the test catches regressions in the actual derivation error:
```scala
assertTrue(err.exists(_.message.contains("CsvFieldEncoder")))
```
And similarly for the non-Product case (line 122): a substring mentioning `Mirror` or `ProductOf` would be a stable anchor.

File: csvzen/core/src/main/scala/com/guizmaii/csvzen/core/CsvConfig.scala [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.guizmaii.csvzen.core
+
+final case class CsvConfig(
+  delimiter: Char = ',',
+  quoteChar: Char = '"',
+  lineTerminator: String = "\r\n",
+)
```
Comment (guizmaii) [inline csvzen/core/src/main/scala/com/guizmaii/csvzen/core/CsvConfig.scala]: **No input validation — several silently-broken configurations are accepted.** Pathological-but-plausible mistakes produce un-parseable CSV with no error at construction time:

- `delimiter == quoteChar` (e.g. both `"`). `writeEscaped` has a branch that accidentally handles this (the `c == quoteChar` check runs before the delimiter check) but the output is then unparseable by any standard reader.
- `delimiter` set to `'\r'`, `'\n'`, or `'"'` — breaks quoting invariants.
- `quoteChar` set to `'\r'` or `'\n'`.
- `lineTerminator` containing characters that aren't `\r` / `\n` (e.g. a stray `;`) — `writeEscaped` only considers `\r`/`\n` when deciding to quote, so field values containing those other characters are emitted verbatim and corrupt the row framing.
- Empty `lineTerminator`.

A `require(...)` block in the companion `apply` (or a smart constructor returning `Either[String, CsvConfig]`) would close these off at construction. It also lets you drop the "pathological config" branch in `FieldEmitter.writeEscaped`.

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+  private val delimiter: Char        = config.delimiter
+  private val quoteChar: Char        = config.quoteChar
+  private val lineTerminator: String = config.lineTerminator
+  private var first: Boolean         = true
+  private val scratch: Array[Char]   = new Array[Char](20)
+
+  private[core] def beginRow(): Unit = first = true
+
+  private[core] def endRow(): Unit = out.write(lineTerminator)
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala:26]: **Partial-row state corruption on mid-row `IOException` is undocumented.** If `out.write(...)` throws inside any `emit*` — e.g. midway through `writeEscaped` (many small writes) — `first` is left in whatever state the partial row reached. A caller that catches the `IOException` and calls `writeRow` again will produce a file with a truncated row glued to a new row, with no leading delimiter on the first new field. Silently invalid CSV.

Two options, pick one:
1. Document on `CsvWriter`/`FieldEmitter` that the writer is unusable after any `IOException` and must be closed.
2. Reset `first = true` inside `endRow()` and add a `poisoned: Boolean` flag set on any caught error; further `emit*` calls throw `IllegalStateException`.

Option 1 is one Scaladoc line and matches the "streaming, no recovery" posture of the library. Happy with either.

**Review by guizmaii — COMMENTED** [review]
Overall a very solid v0.1. CSV escaping is RFC 4180-correct, the hand-rolled `writeInt`/`writeLong` digit conversion is sound (scratch buffer sized correctly at 20 chars, `MIN_VALUE` carve-outs handled), the exhaustive `Short`/`Byte` range tests and 10k-element randomised sweeps for `Int`/`Long` are exactly the right choice for narrow domains, and the codec-vs-escape-hatch byte-equivalence test (`CsvRowEncoderSpec.scala:77`) is a nice touch.

Focused feedback below. Two real test bugs, one missing input validation, and one style-config item. Nothing I'd call a blocker, but the test bugs in particular are worth fixing before the next tag.

**Future API surface** (v0.2, not blocking):
- No public `flush()` on `CsvWriter` — long-running writers can't checkpoint partial output to disk.
- `CsvWriter.open` hard-codes `UTF-8` and default open options (no `APPEND`, no BOM, no `ISO-8859-1`). Worth `charset: Charset = UTF_8, options: OpenOption*`.
- Thread-safety is not documented anywhere. Given the mutable `scratch` buffer and `first` flag on `FieldEmitter`, a single instance is strictly single-threaded — a one-line Scaladoc note on `CsvWriter` would save someone a nasty debug session later.
- No `.scalafmt.conf` in the project, so the 120-col convention from the workspace `CLAUDE.md` is not enforced; three test lines are now over the limit (worst is `CsvFileWriterSpec.scala:8` at 189 chars).

File: core/src/main/scala/com/guizmaii/csvzen/core/CsvRowEncoder.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    out
+  }
+
+  private inline def encodersOf[E <: Tuple]: Array[CsvFieldEncoder[Any]] = {
+    val arr = summonAll[Tuple.Map[E, CsvFieldEncoder]].toArray
+    val out = new Array[CsvFieldEncoder[Any]](arr.length)
+    var i   = 0
+    while (i < arr.length) {
+      out(i) = arr(i).asInstanceOf[CsvFieldEncoder[Any]]
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/CsvRowEncoder.scala:48]: Small note: because `CsvFieldEncoder[-A]` is contravariant, casting `CsvFieldEncoder[X]` to `CsvFieldEncoder[Any]` is a *narrowing* (`CsvFieldEncoder[Any] <: CsvFieldEncoder[X]`), not a widening — i.e. formally unsound by the type system. It's safe in practice because `DerivedCsvRowEncoder.encode` only ever feeds each slot the matching `productElement(i)` value for that position.

Worth a one-line comment here or at the call site so a future reader (or a future refactor that changes the derivation order) doesn't trip over it. Alternatively: keep the array typed as `Array[CsvFieldEncoder[_]]` and do the `asInstanceOf[CsvFieldEncoder[Any]]` only at the single call site in `DerivedCsvRowEncoder`, right next to the `productElement` invariant it relies on.

Related, line 14 of `DerivedCsvRowEncoder.scala`: the `a.asInstanceOf[Product]` is safe because `Mirror.ProductOf` is only synthesized for types that extend `Product`, but it's formally redundant — you can bound `derived[A <: Product]` and drop the cast.

File: build.sbt [hunk 7]
```diff
@@ -3,7 +3,7 @@ import BuildHelper.{noDoc, stdSettings}
 Global / onChangedBuildSource := ReloadOnSourceChanges
 
 ThisBuild / organization := "com.guizmaii"
-ThisBuild / name         := "scala-nimbus-jose-jwt"
+ThisBuild / name         := "csvzen"
```
Comment (guizmaii) [inline build.sbt]: Good catch fixing the `"scala-nimbus-jose-jwt"` leftover, but `name` doesn't actually belong at `ThisBuild` scope at all — it's inherently a per-project setting. The root project's name is already set by `Project(id = "csvzen", ...)` on line 27, and `csvzen-core` overrides it on line 38. Any plugin that reads `(ThisBuild / name).value` would now see "csvzen" for the `csvzen-core` subproject too, which is wrong.

Suggest dropping this line entirely (no replacement needed — the existing `Project(id = ...)` + per-module `name :=` pair covers it).

File: .jvmopts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# This is a comment
+-Dfile.encoding=UTF8
+-Xms128M
+-Xmx4G
+-Xss16M
+-XX:+UseG1GC
+--add-opens=java.base/java.util=ALL-UNNAMED
+--add-opens=java.base/java.lang=ALL-UNNAMED
+--add-exports=java.base/sun.nio.ch=ALL-UNNAMED
```
Comment (guizmaii) [inline .jvmopts]: These four `--add-opens` / `--add-exports` lines look copy-pasted from a Spark/Arrow/Kafka-style project template. csvzen-core only uses `java.io.Writer`, `java.nio.file.Files`, and the `java.time` + `java.util` public APIs — none of which need reflective access to `sun.nio.ch`, `jdk.internal.misc`, `java.base/java.util`, or `java.base/java.lang`.

Suggest dropping:
```
--add-opens=java.base/java.util=ALL-UNNAMED
--add-opens=java.base/java.lang=ALL-UNNAMED
--add-exports=java.base/sun.nio.ch=ALL-UNNAMED
--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED
```
and keeping only `--enable-native-access=ALL-UNNAMED` (if you've verified the JDK25+ warning it silences is actually emitted for this build). Fewer knobs = less surprise when someone later adds a dependency that does need reflection.

**Review by guizmaii — COMMENTED** [review]
Nice turnaround — every inline comment from the previous review has been addressed cleanly:
- `CsvConfig` now validates `delimiter`/`quoteChar`/`lineTerminator` in `require` blocks.
- `CsvWriter` gains `flush()`, a configurable `charset` and `options*`, and a Scaladoc paragraph pinning the thread-safety and "unusable after IOException" stances.
- `FieldEmitter` has a thread-safety note in its Scaladoc.
- `CsvRowEncoder.derived` now bounds `A <: Product` (dropping the `asInstanceOf[Product]` in `DerivedCsvRowEncoder`) and documents the contravariance-narrowing cast in `encodersOf`.
- Unused `ArraySeq` import removed.
- `CsvFieldEncoderSpec` currency test captures into `ok` and asserts on it.
- Negative compile tests pin a message substring in each case.

Small meta-correction: the workspace-level `CLAUDE.md` says 120-col max, but this project's own `.scalafmt.conf` sets `maxColumn = 140`. The project config wins; my previous note about 120 cols was off.

Second-round feedback is now almost entirely about the new build configuration. Two inline items below. Three more that aren't on changed lines of this PR (so noting them here instead):

### Build-config items not inlineable

1. **`.scalafix.conf` exists but `sbt-scalafix` is not in `project/plugins.sbt`** — the scalafix rules (`RemoveUnused`, `OrganizeImports`, `NoAutoTupling`, `RedundantSyntax`, `DisableSyntax` with `noReturns`/`noXml`/`noFinalize`/`noValPatterns`) never actually run. Either add `addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.14.2")` + a `scalafixOnCompile` setting (or a check alias), or delete `.scalafix.conf` so the repo doesn't ship config that looks active but isn't. Silent-no-op config is worse than no config.

2. **Duplicated `organization` in `build.sbt`** — `ThisBuild / organization := "com.guizmaii"` at the top of the file, AND inside the `inThisBuild(...)` block at the bottom. Pick one spot. Convention is to put the whole `inThisBuild(...)` block at the *top* of `build.sbt` (before project definitions) rather than below them, and keep all cross-project defaults there.

3. **`sbt-scoverage` is in `plugins.sbt` but no coverage aliases or `coverageEnabled` settings are defined.** If you're not actively using it, dropping the plugin saves a resolution step per reload. If you plan to use it, add `addCommandAlias("cov", "clean; coverage; test; coverageReport")` or similar.

4. **`sbt-mima-plugin` is absent.** For a library heading to Maven Central via `sbt-ci-release`, a `mimaChecks` gate is worth having even pre-1.0 — catches accidental binary-incompat between 0.1.x → 0.2.x. (Not urgent for this PR.)

File: core/src/main/scala/com/guizmaii/csvzen/core/internal/DerivedCsvRowEncoder.scala [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+final class DerivedCsvRowEncoder[A <: Product](
+  labels: Array[String],
+  encoders: Array[CsvFieldEncoder[Any]],
+) extends CsvRowEncoder[A] {
+
+  val headerNames: IndexedSeq[String] = ArraySeq.unsafeWrapArray(labels)
+
+  def encode(a: A, out: FieldEmitter): Unit = {
+    val n = encoders.length
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/internal/DerivedCsvRowEncoder.scala]: Do we need to recompute `n` each time?

File: core/src/main/scala/com/guizmaii/csvzen/core/CsvWriter.scala [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+  inline def writeAll[A](rows: Iterable[A])(using enc: CsvRowEncoder[A]): Unit = {
+    val it = rows.iterator
+    while (it.hasNext) writeRow(it.next())
+  }
+
+  /** Flushes the underlying `Writer`, pushing buffered bytes to disk. */
+  def flush(): Unit = out.flush()
+
+  override def close(): Unit = out.close()
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/CsvWriter.scala:55]: Can we add a comment, mentioning that `out.close()` also flushed to disk?

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  def emitChar(c: Char): Unit = {
+    fieldPrelude()
+    writeEscapedChar(c)
+  }
+
+  def emitFloat(f: Float): Unit = {
+    fieldPrelude()
+    out.write(java.lang.Float.toString(f))
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala:71]: Should we prefer `String.valueOf` over `Float.toString`? Any diff?

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  def emitDouble(d: Double): Unit = {
+    fieldPrelude()
+    out.write(java.lang.Double.toString(d))
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala:76]: Should we prefer `String.valueOf` over `Double.toString`? Any diff?

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  private def writeInt(i: Int): Unit =
+    if (i == Integer.MIN_VALUE) out.write("-2147483648")
+    else {
+      var n    = i
+      if (n < 0) {
+        out.write('-'.toInt)
+        n = -n
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: Any binary operator to do this which would be more efficient?

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    out.write(java.lang.Float.toString(f))
+  }
+
+  def emitDouble(d: Double): Unit = {
+    fieldPrelude()
+    out.write(java.lang.Double.toString(d))
+  }
+
+  private def writeInt(i: Int): Unit =
+    if (i == Integer.MIN_VALUE) out.write("-2147483648")
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: What about `Integer.MAX_VALUE`?

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+      var n    = i
+      if (n < 0) {
+        out.write('-'.toInt)
+        n = -n
+      }
+      var pos  = scratch.length
+      var cont = true
+      while (cont) {
+        pos -= 1
+        scratch(pos) = ('0' + (n % 10)).toChar
```
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: Any better way to get the `Char` than doing `('0' + ...).toChar`?

File: core/src/test/scala/com/guizmaii/csvzen/core/FieldEmitterSpec.scala [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+      },
+      test("writes negative values") {
+        val values = Seq(-1, -9, -10, -99, -100, -12345, Int.MinValue + 1, Int.MinValue)
+        assertTrue(values.forall(v => firstField(cfg)(_.emitInt(v)) == v.toString))
+      },
+      test("matches Integer.toString on a randomised sweep") {
+        val rng    = new scala.util.Random(0xc0ffee)
+        val sample = (0 until 10000).map(_ => rng.nextInt())
+        assertTrue(sample.forall(v => firstField(cfg)(_.emitInt(v)) == Integer.toString(v)))
+      },
```
Comment (guizmaii) [inline core/src/test/scala/com/guizmaii/csvzen/core/FieldEmitterSpec.scala]: Instead of doing that we can use zio-test PBT capabilities

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 14 above)
Comment (guizmaii) [inline core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: Should we go branchless? What's the benefits of this `if` here?

File: build.sbt (hunk 7 above)
Comment (guizmaii) [reply build.sbt]: Fixed in 2a39220 — dropped both `ThisBuild / name` and the standalone `ThisBuild / organization` (the canonical pair lives in the `inThisBuild` block). Didn't move the block to the top of the file in this commit; happy to do that as a follow-up if you'd like.

File: .jvmopts (hunk 8 above)
Comment (guizmaii) [reply .jvmopts]: Fixed in 2a39220 — dropped the four `--add-opens` / `--add-exports` lines, kept only `--enable-native-access=ALL-UNNAMED`.

File: core/src/main/scala/com/guizmaii/csvzen/core/internal/DerivedCsvRowEncoder.scala (hunk 9 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/internal/DerivedCsvRowEncoder.scala]: Good catch — fixed in 2a39220. Hoisted to a `private val n: Int = encoders.length` on the class so it's computed once at construction time and reused for every row.

File: core/src/main/scala/com/guizmaii/csvzen/core/CsvWriter.scala (hunk 10 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/CsvWriter.scala:55]: Done in 2a39220 — `close()` scaladoc now mentions the flush. While I was at it, made `CsvWriter` extend `java.io.Flushable` (it already exposes `flush()`, so the type just makes that contract explicit alongside the existing `AutoCloseable`).

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 11 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala:71]: No difference. `String.valueOf(f: Float)` is a one-liner in `java.lang.String` that just delegates to `Float.toString(f)` — same call, same allocation, same result. Keeping the explicit `Float.toString` here since we're already in primitive territory and it reads more directly.

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 12 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala:76]: Same answer as for `Float.toString` — `String.valueOf(d: Double)` is just a thin delegate to `java.lang.Double.toString(d)`. No difference at the bytecode level.

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 13 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: Addressed by the refactor in 2a39220 — the explicit `n = -n` is gone. The new code processes digits in the non-positive domain, so the only negation is `n = -n` on the positive-input path, which already compiles to a single `INEG` JVM instruction (= `~n + 1` in two's complement). No bitwise alternative is faster.

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 14 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: `Integer.MAX_VALUE` doesn't need a carve-out — `-Int.MaxValue = -2147483647` fits in `Int` fine. The original `MIN_VALUE` branch was only there because `-Int.MinValue` overflows (no positive Int counterpart in two's complement). The refactor in 2a39220 processes digits in the non-positive domain, sidestepping that overflow entirely, so neither extreme now needs a special case.

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 15 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: `('0' + d).toChar` (or `('0' - d).toChar` in the new non-positive-domain code) compiles to a single `IADD`/`ISUB` plus an `I2C` cast — already minimal. `(0x30 | d).toChar` is bytecode-equivalent (works because `'0' = 0x30` and digits 0–9 fit in the low 4 bits) but not faster on any JIT I know of. A lookup table is strictly slower because of the bounds check + memory load. Keeping the arithmetic form for clarity.

File: core/src/test/scala/com/guizmaii/csvzen/core/FieldEmitterSpec.scala (hunk 16 above)
Comment (guizmaii) [reply core/src/test/scala/com/guizmaii/csvzen/core/FieldEmitterSpec.scala]: Done in 2a39220 — both `emitInt` and `emitLong` random sweeps now use `check(Gen.int) { ... }` / `check(Gen.long) { ... }`. The boundary tests stay as separate cases (`Gen.int` doesn't reliably hit `Int.MinValue`/`Int.MaxValue` in the default sample size, and the boundary cases are the ones that historically broke the digit-conversion code).

File: core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala (hunk 14 above)
Comment (guizmaii) [reply core/src/main/scala/com/guizmaii/csvzen/core/FieldEmitter.scala]: Refactored in 2a39220 — the `MIN_VALUE` branch is gone entirely. The new `writeInt`/`writeLong` process digits in the non-positive domain (`'0' - (n % 10)` instead of `'0' + (n % 10)`), so neither extreme needs a special case. The remaining sign branch (`if (n < 0) write('-') else n = -n`) is just sign emission; going fully branchless there would mean something like `out.write('-' & ((n >> 31) & ...))`, which obscures intent for no measurable win — the branch predictor hits ~100% on whichever path the dominant column is taking.

**Comment by guizmaii** [comment]
### Code review

Found 2 issues:

1. The design doc's algorithm description for `emitInt` / `emitLong` is now stale: it says the implementation special-cases `Int.MinValue` / `Long.MinValue` and "operate[s] on a positive number" after the sign branch, but the implementation does the opposite — it negates positives and operates entirely in the non-positive domain (`'0' - n % 10`) precisely to *avoid* the MIN_VALUE special-case. Commit `b21c260` even added an inline comment explaining the negative-domain trick. Worth updating the spec so it matches the code.

https://github.com/guizmaii-opensource/csvzen/blob/eb397b841395b556ff92dcdc69b942f7dcb60312/docs/superpowers/specs/2026-04-24-csvzen-design.md#L182-L186

2. `sbt-scalafix` is added to `project/plugins.sbt` and `semanticdbEnabled` is set in `build.sbt`, and `.scalafix.conf` defines an active rule set (`DisableSyntax`, `LeakingImplicitClassVal`, `NoAutoTupling`, `NoValInForComprehension`, `RemoveUnused`, `RedundantSyntax`), but nothing actually invokes scalafix — there is no `scalafixOnCompile`, no `scalafixAll` / `fix` / `fmtCheck` command alias, and no CI step. The configured rules never run. Either wire it up (e.g. `ThisBuild / scalafixOnCompile := !insideCI.value` plus a check alias) or drop `.scalafix.conf` and the plugin. Same applies to `sbt-scoverage` (line 4 of `plugins.sbt`) — listed but no `coverageEnabled` / coverage alias defined.

https://github.com/guizmaii-opensource/csvzen/blob/eb397b841395b556ff92dcdc69b942f7dcb60312/build.sbt#L7-L17

https://github.com/guizmaii-opensource/csvzen/blob/eb397b841395b556ff92dcdc69b942f7dcb60312/project/plugins.sbt#L1-L6

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by guizmaii** [comment]
Both addressed in 47828b4.

**Issue 1 (stale design doc):** Updated `docs/superpowers/specs/2026-04-24-csvzen-design.md`. The `emitInt` / `emitLong` paragraph now describes the non-positive-domain digit extraction (`('0' - n % 10)`, no MIN_VALUE carve-out, same trick as JDK `Integer.getChars`). The `scratch` buffer description was also stale (said length 20, justified `-9223372036854775808`); both are corrected to length 19, with the sign-written-separately note.

**Issue 2 (dead lint config):**
- Wired up scalafix: added `ThisBuild / scalafixOnCompile := !insideCI.value` (matches the existing `scalafmtOnCompile` pattern), plus `fix` and `check` command aliases. The CI `formatting` job is now a `lint` job that runs `sbt check` (scalafix + scalafmt + scalafmt.sbt). Verified `sbt check` runs clean on the current tree.
- Dropped `sbt-scoverage` from `project/plugins.sbt` — it was installed without any `coverageEnabled` setting or alias, so it was pure overhead per reload. We can re-add it when there's a concrete reporting destination.
